core/rawdb/ancienttest/testsuite.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package ancienttest
+
+import (
+	"bytes"
+	"crypto/rand"
+	"reflect"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+// TestAncientSuite runs a suite of tests against an ancient database
+// implementation.
+func TestAncientSuite(t *testing.T, New func(kinds []string) ethdb.AncientStore) {
+	// Test basic read methods
+	t.Run("BasicRead", func(t *testing.T) {
+		var (
```

Comment:
This becomes an extremely large function. Perhaps lift the inlined functions out? 
```golang
// TestAncientSuite runs a suite of tests against an ancient database
// implementation.
func TestAncientSuite(t *testing.T, New func(kinds []string) ethdb.AncientStore) {
	// Test basic read methods
	t.Run("BasicRead", testBasicRead)

	// Test batch read method
	t.Run("BatchRead", testBatchRead)
```

Reply:
Also the name `New` as an argument is odd. Perhaps `newFn` or `constructor` or `ctor` ? 
Then the above would become something like 
```
     t.Run("BasicRead", func(t *testing.T ){ testBasicRead ( t, ctor ) } ) 
```

Reply:
Please add this testcase, to trigger the mutability-bug:

```golang
...
	t.Run("NonMutable", func(t *testing.T) { testNonMutable(t, New) })
...


func testNonMutable(t *testing.T, ctor func(kinds []string) ethdb.AncientStore) {
	var (
		db = ctor([]string{"a"})
	)
	defer db.Close()

	// We write 100 zero-bytes to the freezer and immediately mutate the slice
	db.ModifyAncients(func(op ethdb.AncientWriteOp) error {
		data := make([]byte, 100)
		op.AppendRaw("a", uint64(0), data)
		for i := range data {
			data[i] = 0xff
		}
		return nil
	})
	// Now read it.
	data, err := db.Ancient("a", uint64(0))
	if err != nil {
		t.Fatal(err)
	}
	for k, v := range data {
		if v != 0 {
			t.Fatalf("byte %d != 0: %x", k, v)
		}
	}
}
```